File: archive/cdk-py/Pr2086/lib/database_stack.py
"""Database Stack with RDS MySQL Multi-AZ deployment."""

from aws_cdk import CfnOutput, Duration, NestedStack, RemovalPolicy
from aws_cdk import aws_ec2 as ec2
from aws_cdk import aws_rds as rds
from aws_cdk import aws_secretsmanager as secretsmanager
from constructs import Construct


class DatabaseStack(NestedStack):
    """Creates RDS MySQL database with Multi-AZ deployment."""
    
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        vpc: ec2.Vpc,
        db_security_group: ec2.SecurityGroup,
        environment_suffix: str = "dev",
        **kwargs
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)
        
        self.vpc = vpc
        self.db_security_group = db_security_group
        self.environment_suffix = environment_suffix
        
        # Create database
        self._create_database()
    
    def _create_database(self):
        """Create RDS MySQL database with Multi-AZ deployment."""
        
        # Create database credentials in Secrets Manager
        self.db_secret = secretsmanager.Secret(
            self, "prod-db-secret-primary-1",
            description="Database credentials for production web application",
            generate_secret_string=secretsmanager.SecretStringGenerator(
                secret_string_template='{"username": "admin"}',
                generate_string_key="password",
                exclude_characters='"@/\\',
                password_length=32,
            ),
        )
        
        # Create subnet group
        subnet_group = rds.SubnetGroup(
            self, "prod-db-subnet-group-primary-1",
            description="Subnet group for production database",
            vpc=self.vpc,
            vpc_subnets=ec2.SubnetSelection(
                subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS
            ),
        )
        
        # Create parameter group for optimization
        parameter_group = rds.ParameterGroup(
            self, "prod-db-params-primary-1",
            engine=rds.DatabaseInstanceEngine.mysql(
                version=rds.MysqlEngineVersion.VER_8_0_39
            ),
            description="Parameter group for production MySQL database",
            parameters={
                "innodb_buffer_pool_size": "134217728",  # Optimized for t3.micro
                "max_connections": "100",
            },
        )
        
        # Create RDS instance
        self.database = rds.DatabaseInstance(
            self, "prod-database-primary-1",
            engine=rds.DatabaseInstanceEngine.mysql(
                version=rds.MysqlEngineVersion.VER_8_0_39
            ),
            instance_type=ec2.InstanceType.of(
                ec2.InstanceClass.T3, ec2.InstanceSize.MICRO
            ),
            credentials=rds.Credentials.from_secret(self.db_secret),
            database_name="prodwebapp",
            vpc=self.vpc,
            subnet_group=subnet_group,
            security_groups=[self.db_security_group],
            parameter_group=parameter_group,
            multi_az=True,  # High availability
            allocated_storage=20,
            storage_type=rds.StorageType.GP3,
            storage_encrypted=True,  # Encryption at rest
            backup_retention=Duration.days(7),
            deletion_protection=False,  # Allow deletion for testing
            delete_automated_backups=False,
            enable_performance_insights=False,  # Performance Insights not supported for t3.micro
            monitoring_interval=Duration.seconds(60),
            auto_minor_version_upgrade=True,
            removal_policy=RemovalPolicy.DESTROY,  # Allow deletion for testing
        )
        
        # Output database information
        CfnOutput(self, "DatabaseEndpoint", value=self.database.instance_endpoint.hostname)
        CfnOutput(self, "DatabaseSecretArn", value=self.db_secret.secret_arn)

File: archive/cdk-ts/Pr3935/lib/stacks/database-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as rds from 'aws-cdk-lib/aws-rds';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as kms from 'aws-cdk-lib/aws-kms';
import * as secretsmanager from 'aws-cdk-lib/aws-secretsmanager';
import { Construct } from 'constructs';

interface DatabaseStackProps extends cdk.NestedStackProps {
  environmentSuffix: string;
  vpc: ec2.Vpc;
  kmsKey: kms.Key;
  databaseSecurityGroup: ec2.SecurityGroup;
  tags?: { [key: string]: string };
}

export class DatabaseStack extends cdk.NestedStack {
  public readonly databaseSecret: secretsmanager.Secret;
  public readonly database: rds.DatabaseInstance;

  constructor(scope: Construct, id: string, props: DatabaseStackProps) {
    super(scope, id, props);

    // Apply tags to all resources in this stack
    cdk.Tags.of(this).add('iac-rlhf-amazon', 'true');

    // Create database credentials in Secrets Manager
    this.databaseSecret = new secretsmanager.Secret(this, 'DatabaseSecret', {
      secretName: `${props.environmentSuffix}/rds/credentials-v4`,
      description: `RDS database credentials for ${props.environmentSuffix} environment`,
      generateSecretString: {
        secretStringTemplate: JSON.stringify({ username: 'admin' }),
        generateStringKey: 'password',
        excludeCharacters: ' %+~`#$&*()|[]{}:;<>?!\'/@"\\',
        passwordLength: 32,
      },
      encryptionKey: props.kmsKey,
    });

    // Create subnet group for database
    const dbSubnetGroup = new rds.SubnetGroup(this, 'DatabaseSubnetGroup', {
      description: `Database subnet group for ${props.environmentSuffix}`,
      vpc: props.vpc,
      vpcSubnets: {
        subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
      },
    });

    // Create RDS instance with Multi-AZ
    this.database = new rds.DatabaseInstance(this, 'Database', {
      engine: rds.DatabaseInstanceEngine.mysql({
        version: rds.MysqlEngineVersion.VER_8_0_39,
      }),
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.T3,
        ec2.InstanceSize.LARGE
      ),
      vpc: props.vpc,
      subnetGroup: dbSubnetGroup,
      securityGroups: [props.databaseSecurityGroup],
      credentials: rds.Credentials.fromSecret(this.databaseSecret),
      multiAz: true,
      allocatedStorage: 100,
      maxAllocatedStorage: 1000,
      storageType: rds.StorageType.GP3,
      storageEncrypted: true,
      storageEncryptionKey: props.kmsKey,
      monitoringInterval: cdk.Duration.seconds(60),
      enablePerformanceInsights: true,
      performanceInsightEncryptionKey: props.kmsKey,
      backupRetention: cdk.Duration.days(7),
      preferredBackupWindow: '03:00-04:00',
      preferredMaintenanceWindow: 'sun:04:00-sun:05:00',
      deletionProtection: true,
      removalPolicy: cdk.RemovalPolicy.SNAPSHOT,
    });
  }
}
